modules/ding_gatewayf/ding_gatewayf.module

```diff
@@ -0,0 +1,458 @@
+            )), 'warning');
+          }
+        }
+        else {
+          // Set new default message.
+          drupal_set_message(t('Something went wrong. Please try again or !url of WAYF.', array('!url' => l(t('logout'), DING_GATEWAYF_LOGOUT_URL))), 'warning');
+        }
+      }
+      elseif (_ding_gatewayf_registration_is_registration_request()) {
+        // Account exists in the library system, so skip self registration
+        // request.
+        drupal_set_message(t('You already have an user account at the library. So we have just logged you in.'));
+
+        // Remove the redirect to acceptance page.
+        unset($_REQUEST['destination']);
+        unset($_GET['destination']);
+
+        // Redirect to the front-page.
+        _ding_gatewayf_redirect_user('<front>');
+      }
+    }
+    else {
+      drupal_set_message(t('Something went wrong. Please try again or !url of WAYF.', array('!url' => l(t('logout'), DING_GATEWAYF_LOGOUT_URL))), 'warning');
+    }
+  }
+
+  _ding_gatewayf_redirect_user();
+}
+
+/**
+ * Redirect the user after login.
+ *
+ * If destination query string exists the user is redirect to it else send to
+ * profile page.
+ *
+ * @param string $url
+ *   The url to redirect to if destination is not set in the request. Defaults
+ *   to /user.
+ */
+function _ding_gatewayf_redirect_user($url = 'user') {
+  if (!empty($_REQUEST['destination'])) {
+    // When logged in with gatewayf the destination is part of the get request.
+    // So to use the cleaned up destination we unset it.
+    unset($_GET['destination']);
+
+    // We do not user drupal_get_destination here as it will set the destination
+    // if one do not exists.
+    $url = $_REQUEST['destination'];
+  }
+
+  drupal_goto($url);
+}
+
+/**
+ * Get user credentials from SAML attributes and the user provider.
+ *
+ * @param array $attributes
+ *   WAYF attributes.
+ *
+ * @return array
+ *   Attributes filter based on the active provider. Defaults to
+ *   'schacPersonalUniqueID' with social security number (CPR).
+ */
+function _ding_gatewayf_get_credentials(array $attributes) {
+  // Set default required attributes.
+  $required_attributes = array(
+    'schacPersonalUniqueID' => array(
+      'field' => 'CPR',
+      'authname' => TRUE,
+    ),
+  );
+
+  // Ask the provider about required attributes and give it a change to override
+  // defaults.
+  if (ding_provider_implements('wayf', 'attributes')) {
+    $required_attributes = ding_provider_invoke('wayf', 'attributes', $required_attributes);
+  }
+
+  // Filter the attributes from WAYF/SAML and only return required once.
+  return _ding_gatewayf_get_required_attributes($attributes, $required_attributes);
+}
+
+/**
+ * Extract required attributes from available WAYF attributes.
+ *
+ * @param array $attributes
+ *   WAYF attributes.
+ * @param array $required_attributes
+ *   Required attributes.
+ *
+ * @return array
+ *   Will be empty if not all required attributes is present.
```

Comment:
It says: "Will be empty if not all required attributes is present." I cannot see how the implementation will return an empty array if there are two entries in `$required_attributes` and `$attributes` only contain of them. Instead it will return an array containing one entry.